Misc type fixes and enhancements I've accumulated

SnaveSutit opened the pull request:
## Notes
- The commit called "Add `Texture._static` types" looks scary because A: Some lines got moved around in the file between our repos somehow, and B: I had to wrap the file in `declare global { ... }` to allow importing `FSWatcher`. As adding imports to change the file's default context from global to module.

    The important changes made to `types/textures.d.ts` in that commit (other than the above changes) are as follows:
    - ### Make `Texture.material` a get / set pair.
        Several of my plugins, including Animated Java, need to modify this property, and it's actual implementation [here](https://github.com/JannisX11/blockbench/blob/368efc7c8275d11fac355efa90720ebcd850f3b8/js/texturing/textures.js#L335) includes a setter.
        ```diff
        - readonly material: THREE.ShaderMaterial
        + get material(): THREE.ShaderMaterial
        + set material(THREE.ShaderMaterial)
        ```
    - ### Added `Texture.loadContentFromPath()`
        ```diff
       +/**
		+ * Loads file content **only**.
		+ *
		+ * Does not read `png.mcmeta`, or attempt to overwrite an existing texture in the project with the same name.
		+ *
		+ * Used internally when loading `.bbmodel` files
		+ * @param path
		+ */
		+loadContentFromPath(path: string): this
	        ```

Please open a discussion if there's anything I've changed here that seems wrong, or that you disagree with.

SnaveSutit committed:
Remove unnecessary Partial

All of the options in CubeFaceOptions are already optional.

SnaveSutit committed:
Uncomment V3_set

V3_set is a valid overload, and should be included in the types

SnaveSutit committed:
🚧 Group and texture changes

SnaveSutit committed:
Plugins type fixes

SnaveSutit committed:
🛠️ Add marked functions

SnaveSutit committed:
Add `Texture._static` types

SnaveSutit committed:
Add basic `SharedAction` types

SnaveSutit committed:
Add Mode constructor overload, and add missing ModeOption properties

SnaveSutit committed:
Make parameter `position` of `Toolbar.add` optional

SnaveSutit committed:
Add `name` parameter to `MenuSeparator` constructor

SnaveSutit committed:
Make all properties of `ICubeOptions` optional, instead of using `Partial` every time it's used.

It makes more sense to default to the most common use-case, than to modify the type every time it's used.

If someone really needs a version of `ICubeOptions` with entirely required properties, they can use `Required<ICubeOptions>`.

SnaveSutit committed:
Make `Texture.mode` type `never` to keep the depreciated message

This will better help redirect developers using `mode` to the new `Texture.internal` value by showing them a depreciated warning that links to the new property instead of just complaining that `mode` doesn't exist.

SnaveSutit committed:
Auto-format with Prettier on save in VSCode

SnaveSutit committed:
Prettier run on whole repo

✨ Standardization ✨

SnaveSutit committed:
`Collection.all` is a getter, not a function

JannisX11 commented on types/cube.d.ts:
```diff
@@ -2,27 +2,27 @@
-	uv_offset: ArrayVector2
-	faces: Partial<Record<CardinalDirection, CubeFaceOptions>>
+	uv_offset?: ArrayVector2
+	faces?: Record<CardinalDirection, CubeFaceOptions>
```
The Partial here is required because it otherwise expects all face directions to be present.

SnaveSutit replied:
Damn, you're right.. How did I miss that :facepalm:

SnaveSutit committed:
Improve types for `CustomElements.SelectInput`

JannisX11 commented on types/action.d.ts:
```diff
@@ -110,19 +118,19 @@ declare class Keybind {
 
 declare class MenuSeparator {
-	constructor()
+	constructor(name?: string)
```
This actually supports two optional string arguments, id and label.

JannisX11 commented on types/global.d.ts:
```diff
@@ -132,7 +132,7 @@ declare global {
 		V3_set(x: number, y: number, z: number): this
 		V3_set(values: ArrayVector3): this
-		//V3_set(value: THREE.Vector3): this
+		V3_set(value: THREE.Vector3): this
```
We talked about on Discord that this wasn't supported. I've added it in a commit now for 4.12.3, but until that's out let's keep it commented.

SnaveSutit replied:
Whoops, that was unintentional. Probably a side-effect of branch divergence

JannisX11 commented on types/plugin.d.ts:
```diff
@@ -97,6 +97,8 @@ declare class BBPlugin {
 	disabled: boolean
 	title: string
 	author: string
+	id: string
```
These are duplicates

SnaveSutit replied:
Looks like the merging process didn't go as smoothly as I'd like. I wonder why VSCode didn't yell at me until I opened the file, I have workspace error checking enabled :thinking:

SnaveSutit committed:
Parameter `key for `Keybind.getKey` is optional.

JannisX11 commented on types/plugin.d.ts:
```diff
@@ -144,6 +146,17 @@ declare class BBPlugin {
+}
+
+declare class Plugins {
+	static all: BBPlugin[]
```
These are duplicates

JannisX11 commented on types/textures.d.ts:
```diff
@@ -1,456 +1,481 @@
+		readonly display_height: number
+		readonly ratio: number
+		static selected?: Texture
+		readonly _static: {
```
Unsure about this. _static is not really meant to be used by plugins, and the properties that are supposed to be accessible (material and selection) are intentionally accessible via getter. I guess there is no harm in having them in here in theory, except that there will be no effort to keep these working across updates.
Are you using any static properties directly for anything?

SnaveSutit replied:
Hmm, This comment made me do a double take when I went to check on my code that accesses `_static.material`.

I was performing some extreme texture shenanigans today, but the code appears to have completely moved away from needing these two functions, as they're exactly identical to the inherited functions.

![image](https://github.com/user-attachments/assets/a5598a5d-af1d-4691-bba4-ec170585ce29)

I will remove these for now.

JannisX11 reviewed (commented):
Nice! Added some notes.

SnaveSutit committed:
Remove `Texture._static` types

SnaveSutit committed:
Correct `MenuSeparator` constructor parameters.

SnaveSutit committed:
Make `ICubeOptions.faces` Partial<>

SnaveSutit committed:
Remove duplicate declarations in `BBPlugin` and of `Plugin`

SnaveSutit committed:
Re-comment `Array.V3_set`

SnaveSutit commented:
Awesome notes, thank you!
_I would make a horrendous QA checker..._ 😆 

Also, thank you for introducing me to the Github comments system!
I have no idea how I've never heard of such a useful feature before...

SnaveSutit committed:
Remove unnecessary global declarations.

From what I have tested, these modules export their own global declarations. Attempting to re-declare them is throwing "Cannot redeclare block-scoped global" errors in TypeScript.

SnaveSutit committed:
Import `tinycolor2`, and wrap types in global scope

You'd think we could get better diffs for this kind of change. Oh well.

SnaveSutit committed:
Add `uuid` property to `TextureGroupOptions`

SnaveSutit committed:
Add static property `all` to `TextureGroup`

SnaveSutit committed:
Revert "Add `uuid` property to `TextureGroupOptions`"

This reverts commit 40873502db4e00fdad79245196691708b9911603.

SnaveSutit committed:
Rename `TextureGroupData` -> `TextureGroupOptions`

All the other constructor "data" interface are using the suffix "Options".

SnaveSutit committed:
Re-Add `uuid` property to `TextureGroupOptions`

I thought I was wrong to include the `uuid` in `TextureGroupOptions` because the `TextureGroup` constructor doesn't use the `uuid` in the provided options. However, the `getSaveCopy` and `OutlinerElement.fromSave` functions do take `uuid` into account. So it's still important to have.

I've used some type trickery to adjust the return types of all the functions that deal with `TextureGroupOptions`-like data.

SnaveSutit committed:
Make `getSaveCopy` generic as a temporary solution

SnaveSutit committed:
Bump version to 4.12.2

SnaveSutit committed:
Change return type of `Array.V3` functions to `ArrayVector3`

Makes it easier to use them, as you don't need to include `as ArrayVector3` after every call.

SnaveSutit committed:
Add `menu` and `file_menu` Animation properties.

SnaveSutit committed:
Add `structure` property to `MenuBar`

SnaveSutit committed:
Add warning for TypeScript users attempting to use `_Animation` as a run-time value.

SnaveSutit committed:
Add `setName` method to `Action`

JannisX11 commented:
> Also, thank you for introducing me to the Github comments system! I have no idea how I've never heard of such a useful feature before...

It's quite cool. With the Github Pull Request extension you can also directly review a PR and add comments in VS Code, and they'll sync with the browser.

JannisX11 committed:
Remove stray type from MenuBar
